src/components/sketches/Scene5_happy.ts

```diff
@@ -0,0 +1,266 @@
+    // Max number of ringParticles
+    p.frameRate(30);
+  };
+
+  p.windowResized = () => {
+    p.resizeCanvas(window.innerWidth, window.innerHeight, true);
+  };
+
+  p.draw = () => {
+    p.background(0, 0, 0, 25); // easy trails but deletes rings quickly
+
+    p.ellipse(p.mouseX,p.mouseY, 5,5);
+    if (p.random(1) < 0.03) {
+      fireworks.push(new Firework());
+    }
+
+    for (let i = fireworks.length - 1; i >= 0; i--) {
+      fireworks[i].update();
+      fireworks[i].show();
+
+      if (fireworks[i].done()) {
+        fireworks.splice(i, 1);
+      }
+    }
+
+    // If too many ringParticles on screen
+    if (ringParticles.length > MAX_RING_PARTICLES) {
+      // Delete old ringParticles (from beginning of array)
+      ringParticles.splice(0, p.floor((ringParticles.length - MAX_RING_PARTICLES * 0.8) * EASING));
+    }
+
+    // Loop through the array and show each particle
+    for (let i = 0; i < ringParticles.length; i++) {
+      ringParticles[i].update();
+      if (ringParticles[i].delete()) {
+        ringParticles.splice(i, 1);
+      } else {
+        ringParticles[i].show();
+      }
+    }
+  };
+  // Spawn ringParticles in ring with centre mouse x and y coords
+  p.mouseClicked = () => {
+    for (let i = 0; i < RING_COVERAGE; i += 4) {
+      const c = new RingParticle(p.mouseX, p.mouseY, i);
+      ringParticles.push(c);
+    }
+  };
+
+  class Firework {
+    public explodeSize: number;
+    public exploded: boolean;
+    private _firingParticle: Particle;
+    private _particles: Array<Particle>;
+
+    constructor() {
+      this.explodeSize = p.random(0, EXPLODE_RANGE);
+      this._firingParticle = new Particle(p.random(p.width), p.height, true);
+      this.exploded = false;
+      this._particles = [];
+    }
+
+    update() {
+      if (!this.exploded) {
+        this._firingParticle.update(gravity);
+
+        if (this._firingParticle.vel.y >= 0) {
+          this.explode();
+        }
+      }
+
+      for (let i = this._particles.length - 1; i >= 0; i--) {
+        this._particles[i].update(gravity);
+
+        if (this._particles[i].done()) {
+          this._particles.splice(i, 1);
+        }
+      }
+    }
+
+    done() {
+      return (this.exploded && this._particles.length === 0);
+    }
+
+    show() {
+      if (!this.exploded) {
+        this._firingParticle.show();
+      }
+
+      for (let i = 0; i < this._particles.length; i++) {
+        this._particles[i].show();
+      }
+    }
+
+    explode() {
+      this.exploded = true;
+      for (let i = 0; i < this.explodeSize * EXPLODE_OFFSET; i++) {
+        const particle = new Particle(
+          this._firingParticle.pos.x,
+          this._firingParticle.pos.y,
+          false,
+        );
+        this._particles.push(particle);
+      }
+    }
+  }
+
+  class Particle {
+    pos: p5.Vector;
+    public isFirework: boolean;
+    lifespan: number;
+    private _acc: p5.Vector;
+    rand: number;
+    vel: p5.Vector;
```

Comment:
these look good!

generally my rule of thumb in deciding public vs private is the following:
* should the member variable be accessible to other functions? `public`
  * **note** if you want the variable to be accessible but IMMUTABLE then do the following:
      ```
      public readonly var: <type>;
      ``` 
* else `private`

Reply:
is my ring particle class good?

also in this bit, _acc and isFirework are both accessed by class functions within Particle. None of these items are being accessed by other functions outside the class... so how could one be priv and one be public?

Reply:
mmmm the way i think about it is even though it isn't being accessed ATM. If you could make something public, then you would denote it as public and if you should NEVER have access to something, then it would be private

Reply:
ring particle class is good!

Reply:
gotcha